docs/proposals/20190708-MachineConfig.md

```diff
@@ -0,0 +1,293 @@
+install:
+  wipe: false
+  force: true
+  boot:
+    device: /dev/sda
+    size: 1024000000
+    kernel: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/vmlinuz
+    initramfs: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/initramfs.xz
+  root:
+    device: /dev/sda
+    size: 2048000000
+    rootfs: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/rootfs.tar.gz
+  data:
+    device: /dev/sda
+    size: 4096000000
+```
+
+#### Worker:
+```
+#!talos
+version: ""
+security: null
+services:
+  init:
+    cni: flannel
+  kubeadm:
+    configuration: |
+      apiVersion: kubeadm.k8s.io/v1beta1
+      kind: JoinConfiguration
+      discovery:
+        bootstrapToken:
+          token: '{{ .KubeadmTokens.BootstrapToken }}'
+          unsafeSkipCAVerification: true
+          apiServerEndpoint: {{ index .MasterIPs 0 }}:443
+      nodeRegistration:
+        taints: []
+        kubeletExtraArgs:
+          node-labels: ""
+          feature-gates: ExperimentalCriticalPodAnnotation=true
+  trustd:
+    token: '{{ .TrustdInfo.Token }}'
+    endpoints: [ {{ .Endpoints }} ]
+networking:
+  os:
+    devices:
+      - interface: lo
+        cidr: x.x.x.x/32
+      - interface: eth0
+        dhcp: true
+install:
+  wipe: false
+  force: true
+  boot:
+    device: /dev/sda
+    size: 1024000000
+    kernel: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/vmlinuz
+    initramfs: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/initramfs.xz
+  root:
+    device: /dev/sda
+    size: 2048000000
+    rootfs: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/rootfs.tar.gz
+  data:
+    device: /dev/sda
+    size: 4096000000
+```
+
+
+## Proposal
+
+As you can see, there are quite a few fields that are similar among all node types, while the main difference between them is the kubeadm templates. Given all the proper info, we should be able to generate the kubeadm configuration on the fly depending on node type and abstract the need to know all the different options away. I'm proposing a machine configuration that looks like this:
+
+
+```
+#!talos
+version: "v0.1.0"
+security:
+  os:
+    ca:
+      crt: "{{ .Certs.OsCert }}"
+      key: "{{ .Certs.OsKey }}"
+    token: abc.1234
+  kubernetes:
+    ca:
+      crt: "{{ .Certs.K8sCert }}"
+      key: "{{ .Certs.K8sKey }}"
+    token: abc.1234
+machine:
+  nodeType: init, controlplane, worker
+  masters: [x.x.x.x, y.y.y.y, z.z.z.z]
+  kubeletExtraArgs: {}
+cluster:
+  clusterName: {{ .ClusterName }}
+  controlPlane:
+    apiServerExtraArgs: {}
+    controllerManagerExtraArgs: {}
+    schedulerExtraArgs: {}
+    networking:
+      dnsDomain: {{ .ServiceDomain }}
+      podSubnet: {{ index .PodNet 0 }}
+      serviceSubnet: {{ index .ServiceNet 0 }}
+networking:
+  os:
+    devices:
+      - interface: lo
+        cidr: x.x.x.x/32
+      - interface: eth0
+        dhcp: true
+install:
+  wipe: false
+  force: true
+  boot:
+    device: /dev/sda
```

Comment:
When multiple volume types specify the same device, as in the case here, does the system just understand that the disk should be partitioned out in some way?  If wipe is false and there is an existing partition table, what happens?

Reply:
We have some intelligence in the system that knows how to partition the block device if `boot` and `data` specify the same device. If wipe is false and there is an existing partition table, Talos will write over that partition table regardless. `wipe` basically indicates that the disk should be zeroed out before partitioning.

Reply:
That still seems to be bit wonky.  Why would you zero data at the top level (and not at the volume level)?  I could understand that from a data security perspective, but that seems to be rather out-of-scope here.

Reply:
That is essentially it, purely for security purposes. It has zero bearing on the Talos functionality so I agree that it is out of scope.